pkg/util/sys/device.go

```diff
@@ -399,17 +409,83 @@ func parseUdevInfo(output string) map[string]string {
-		return false, "", fmt.Errorf("failed to determine if the device %q is available. %v", devicePath, err)
+		return false, "", errors.Wrapf(err, "failed to determine if device %q is available", devicePath)
+	}
+	items := strings.Fields(output)
+	if len(items) != 4 {
+		return false, "", errors.Errorf(
+			"failed to determine if device %q is available. lsblk output should have 4 fields but has %d: %s", devicePath, len(items), items)
 	}
+	rm := items[0]
+	ro := items[1]
+	sizeStr := items[2]
+	devType := items[3]
 
-	if CVInventory.Available {
-		return true, "", nil
+	rejectedList := []string{}
+
+	// must not be removable
+	if rm != "0" { // 0 means not removable, 1 means removable
+		rejectedList = append(rejectedList, "removable")
+	}
+
+	// must not be read-only
+	if ro != "0" { // 0 means not read-only, 1 means read-only
+		rejectedList = append(rejectedList, "read-only")
+	}
+
+	// must be large enough
+	sizeBytes, err := strconv.Atoi(sizeStr)
+	if err != nil {
+		return false, "", errors.Wrapf(err,
+			"failed to determine if device %q is available. failed to interpret reported size %q", devicePath, sizeStr)
 	}
+	if sizeBytes < 5368709120 /* 5368709120 = 5GB */ {
+		rejectedList = append(rejectedList, "insufficient space (< 5GB)")
+	}
+
+	// must be raw device or partition
+	// rejects devType == "lvm" if is an LVM device
+	if devType != DiskType && devType != PartType {
+		rejectedList = append(rejectedList, fmt.Sprintf(`device type %q is not acceptable; should be raw device ("disk") or partition ("part")`, devType))
+	}
+
+	// must not be locked
+	// - try to open in read-write exclusive mode to test if it's locked or not
+	f, err := os.OpenFile(devicePath, os.O_RDWR|os.O_EXCL, os.FileMode(0))
+	if err != nil {
+		rejectedList = append(rejectedList, "locked")
+	}
+	err = f.Close()
+	if err != nil {
+		// failure to close the file pointer will cause problems later
+		return false, "", errors.Wrapf(err, "failed to close device %q when checking for a device lock", devicePath)
+	}
+
+	// must not already be a bluestore OSD
+	// ceph-bluestore-tool show-label returns EXIT_FAILURE on failure which is usually 1 in linux but could be different????
+	// TODO: do we need to handle different translations of ceph-bluestore-tool?
+	output, err = executor.ExecuteCommandWithCombinedOutput("ceph-bluestore-tool", "show-label", "--dev", devicePath)
+	if err != nil {
+		availableText := "unable to read label"
+		// error, plus the availableText indicates the device was found by ceph-bluestore-tool and
+		// it's not a bluestore OSD. error without the availableText is some other error, and we
+		// shouldn't assume the device is available
+		if !strings.Contains(output, availableText) {
+			return false, "", errors.Wrapf(err, "failed to determine if device %q is available (looking for %q): %q", devicePath, availableText, output)
+		}
+	} else {
+		// is a bluestore OSD
+		rejectedList = append(rejectedList, "has BlueStore device label")
+	}
```

Comment:
@leseb what do you think here?

Reply:
I think it's ok but also see my comment above.